### ISSUE TITLE
Add device names

### ISSUE THREAD

**Issue by thorpelawrence** [thorpelawrence/alexa-spotify-connect#1]
Instead of using numbers for devices, being able to assign devices names would make it easier to remember which is which

**Comment by RobPierce** [comment]
I was going to have a look at implementing the device names logic and I'm thinking it will need a custom slot type - reading through the documentation and SO it seems that if you define a custom slot type with at least one entry then alexa will still pass through other spoken terms if it can recognise them so I was thinking of defining a custom slot with some standard room names and then hopefully if users have other names for their devices they should still get through to then compare to the device names.

If I put together the logic for processing the device names, would you be able to create the custom slot type (I'm guessing it needs to be on your amazon developer portal?) and test if it works?

**Comment by thorpelawrence** [comment]
I'd be happy to do that, I've actually had a go at implementing the slot but never gotten around to finishing it, check the `device-slots` branch if you're interested

**Comment by RobPierce** [comment]
That looks pretty good, is it not working or has it just not been tested?

**Comment by thorpelawrence** [comment]
I haven't gotten a chance to test it yet, but it should be mostly complete, although it might need a bit of tweaking

**Comment by RobPierce** [comment]
Sounds good, will take a look through but at first glance I can see anything missing.

**Comment by RobPierce** [comment]
All looks like it should work in theory - one thing I've noticed is that when you list devices, it will only list spotify connect devices that are online since that's what the API returns but on my spotify windows application, it often lists some of my other devices that are in standby (I have various yamaha music cast devices that are always "connected" but not on most of the time) and if I select one of those, it will turn it on and play music on it - I wonder if the windows app caches device data and will somehow turn on a device if it's available but in standby mode, if so it could be neat to try storing device lists between sessions (I think you can use Amazon DynamoDB) so that it's possible to connect to devices on your network that are in standby?

Sorry, without creating my own skill with this code in order to test it I can only speculate!

**Comment by RobPierce** [comment]
Nope, ignore my comment about storing device ids - there must be something else that the spotify windows app does to log into those devices when in standby as I just tried using their api to connect to a device that wasn't on the list of returned devices and it returns a 404 :(

**Comment by thorpelawrence** [comment]
That's interesting, I wasn't aware of that, I used to cache devices for 10 minutes so you could use the same numbers without needing to list again.


I will have a look at setting up a development instance of the app to try it out, as it would be a big change to make in production. I may be able to get some beta testers so let me know if you're interested.

**Comment by RobPierce** [comment]
Would definitely be interested in beta testing.

**Comment by thorpelawrence** [comment]
I've got a beta version running, I just need an email address to add to beta testing- if you'd rather not post it here feel free to send me an email (link on my site)

**Comment by ryanlaux** [comment]
Can i be added to beta testing as well? Would you need any email or amazon acct?

**Comment by thorpelawrence** [comment]
@ryanlaux I would need the email address associated with your Amazon account

**Comment by ryanlaux** [comment]
Cool, I sent you an email to the one on your profile

**Comment by Fredde87** [comment]
I too would be very interested in this feature. If you are still doing beta tests then I would be happy to be a tester too.

**Comment by thorpelawrence** [comment]
I'd be happy to add you to the beta if you send me an email, but be aware that at the moment it is still running the old version, and is not guaranteed to be working correctly as I haven't been able to work on it recently.

**Comment by Fredde87** [comment]
Hi Lawrence,

No worries, I am away with work for a month so probably wont be able to
test it until I get back home. My email address is
fredrik.ringertz@gmail.com


Best Regards

Fredrik

**Comment by xaphod** [comment]
Hi - i have a Denon AVR with Spotify Connect and I can *sometimes* get this skill working with it. Biggest issues: the ordering of devices changes often (my Denon flips between device 2 or 3, when the number of devices on the account is still 3), and sometimes the Denon just disappears (isn't on the list anymore).
If I can be helpful in testing let me know.

**Comment by xaphod** [comment]
OK I fixed it... for me. Here is what I did on my fork, here: https://github.com/xaphod/alexa-spotify-connect

pre: renamed my Denon AVR to "Living Room"
1. Refactored playing/transferring code to look up the devices each time the `.../me/player/...` Spotify endpoints are hit
2. Got rid of device selection/listing: now it just uses a hardcoded device, which is "Living Room" in my fork
3. Removed the cache since we're no longer using it
4. Updated samples / utterances to include "stop" because I don't think of the music stopping as "pausing" :)

It took me a fair while to figure out how to do all this. Some things you could do in your repo to help the next person:
A) add the Alexa skill's JSON to the repo. I've done this on my fork: see `skill/skill.json`
B) mention in README that you need to make a Spotify developer account, and most importantly, list the detailed info you need to enter on the Alexa developer console to get the auth working. **I've entered this in my fork's README** if you want to copy/paste :)
C) mention in README that you need to host the app at somewhere like Heroku, and that the way you've got it now would mean that the URL to set for the Endpoint on the Alexa developer console is https://somewhere.in.the.cloud/**connect** -- I don't know Express so this took me a while to figure out

Thanks a million billion for making this Skill, it was super easy to update your code because you have lots of great comments in it. Thanks again!

**Comment by thorpelawrence** [comment]
Hi @xaphod,

Glad to hear you got it working, feel free to add whatever changes you want to the readme in a PR, it would be good to do it from a third party perspective as I don't remember that well because it was quite a long time ago

As far as generating the skill JSON `skill/skill.js` should generate the JSON needed for adding to the development site; I believe Amazon may have updated their developer site so this might need some small changes but should still work

The `/connect` part is from the dependency [alexa-app](https://github.com/alexa-js/alexa-app) as far as I'm aware and is based on the skill name (i.e. mine is _connect_)

I actually planned to have settings because it would be a nice way to manage things, but there is no 'official' support for this other than storing them on the server and accessing them through voice commands, which I'm sure you can see could get confusing

**Comment by xaphod** [comment]
@sailz86 I haven't published it because it is hardcoded to "Living Room" and uses "Living Room" as the Alexa invocation, which is probably not allowed. So you'd have to host your own web service etc - lot of work if you've not done this before. I think it is not worth publishing until somehow the user can specify the name of the device -- ie. the web service would have to use some persistent storage like Firebase... I don't see myself having time for doing that

**Comment by thorpelawrence** [comment]
Added the development steps to the readme, thanks for your help @xaphod

**Comment by NicoWeio** [comment]
I just implemented my own version of this skill, and my implementation of device names works flawlessly! If you don't mind, I'll go ahead and post it here.

```
const Fuse = require('fuse.js');
const PlayOnDeviceIntentHandler = {
  canHandle: Helpers.canHandleIntent('PlayOnDeviceIntent'),
  async handle(handlerInput) {
    let deviceQuery = handlerInput.requestEnvelope.request.intent.slots.Device.value;
    let s = Spotify(handlerInput);
    const devices = (await s.getMyDevices()).body.devices;
    const results = (new Fuse(devices, {
      keys: ['name'],
      threshold: 0.4, //TODO adjust to a sensible value
    })).search(deviceQuery);

    if (results.length) {
      let selectedDevice = results[0].item;

      await Spotify(handlerInput).transferMyPlayback({
        deviceIds: [selectedDevice.id],
        play: true
      });

      return handlerInput.responseBuilder
        .speak(`Okay, spiele auf ${Helpers.escapeContent(selectedDevice.name)}`)
        .getResponse();
    } else {
      return handlerInput.responseBuilder
        .speak('Ich habe kein passendes Gerät gefunden. Versuche es bitte noch einmal.')
        .reprompt('Versuche es noch einmal. Ich benötige den Namen des Geräts, auf dem du Musik abspielen möchtest.')
        .getResponse();
    }
  }
};
```
As you can see, I used **fues.js** to do fuzzy matching and [spotify-web-api-node](https://github.com/thelinmichael/spotify-web-api-node) just for comfort. 
What do you think? Do you this a sensible way to implement it?

**Comment by thorpelawrence** [comment]
@NicoWeio looks great! Were you able to get it to work with any word for the device name? I think I was having problems with that but it was a couple of years since I looked - also I can't remember how I was trying to do the fuzzy search but fuse.js looks good and great you got it to work

Is this something you would be happy to share as a PR for this repo maybe?

**Comment by NicoWeio** [comment]
Oops, I totally forgot about that part in my previous comment: I use the rather new (2018) _AMAZON.SearchQuery_ slot type, which works fine for me as a German speaker (available in all locales AFAIK). I didn't conduct a lot of tests on how good Alexa understands device names, but it appears to be quite reliable. 

I guess this skill's architecture isn't too complicated, so I'll try and make a PR in the near future.

**Comment by thorpelawrence** [comment]
> Oops, I totally forgot about that part in my previous comment: I use the rather new (2018) _AMAZON.SearchQuery_ slot type, which works fine for me as a German speaker (available in all locales AFAIK). I didn't conduct a lot of tests on how good Alexa understands device names, but it appears to be quite reliable.
> 
> I guess this skill's architecture isn't too complicated, so I'll try and make a PR in the near future.

That's good to know, at the time I was starting there was `AMAZON.LITERAL` but it wasn't being supported anymore (I think they removed it because of privacy reasons or skills being able to record voice forever or something), I think there was `DEVICES_ROOMS` I was trying to use but that was quite limited. `AMAZON.SearchQuery` looks good for this though!

That would be great, thank you!

**Comment by NicoWeio** [comment]
For some reason, I can't get the skill to run locally. I cloned the repo, installed the dependencies, and ran `npm start`, i.e. `node connect.js`. I use [ngrok](https://ngrok.com/) to register my local instance as endpoint in the Developer Console. However, all requests fail with HTTP 404, where express says _Cannot POST_. Do you have an idea what went wrong?
Another thing I noticed is how _DevicePlayIntent_ and _DeviceTransferIntent_ are somewhat "entangled". I wonder if we could use [play](https://developer.spotify.com/documentation/web-api/reference/player/start-a-users-playback/) for both triggering play and transferring playback. Then we could merge these two intents. 
If you'd like, you can take a look at [my skill](https://github.com/NicoWeio/Alexa-Spotify-Remote). Apart from a few features, some error handlers missing, and the lack of a database, it works just like yours.

**Comment by thorpelawrence** [comment]
I haven't had any experience testing with ngrok unfortunately, is it maybe a firewall issue or something? Is the issue POSTing to the Spotify API? Maybe an access token issue? I think I agree, the only difference I can see between play/transfer is that transfer won't automatically play if if was not already playing - maybe people will use this?

**Comment by NicoWeio** [comment]
> Is the issue POSTing to the Spotify API?

No, it was about requests from Alexa. Nevermind, I used the wrong endpoint, `/` instead of `/connect`.

> I think I agree, the only difference I can see between play/transfer is that transfer won't automatically play if if was not already playing - maybe people will use this?

Well, maybe. I can only say that _I_ wouldn't use this, because then: why should I transfer in the first place?

**Comment by NicoWeio** [comment]
I've just opened PR #69. 🙃

**Comment by thorpelawrence** [comment]
Hi all, looks like Amazon have approved the changes, so the update should be live, thanks for the help

**Comment by NicoWeio** [comment]
Great, thanks for merging! :)
